packages/react-components/react-jsx-runtime/src/jsx.ts

```diff
@@ -0,0 +1,89 @@
+import * as React from 'react';
+import {
+  SlotRenderFunction,
+  UnknownSlotProps,
+  SLOT_EXTERNAL_CHILDREN_SYMBOL,
+  SLOT_INTERNAL_CHILDREN_SYMBOL,
+} from '@fluentui/react-utilities';
+
+type WithMetadata<Props extends {}> = Props & {
+  [SLOT_EXTERNAL_CHILDREN_SYMBOL]: React.ReactNode | SlotRenderFunction<Props>;
+  [SLOT_INTERNAL_CHILDREN_SYMBOL]: React.ReactNode | SlotRenderFunction<Props>;
+};
+
+/**
+ * Equivalent to React.createElement but supporting v9 slot API
+ */
+export function jsx<P extends {}>(
+  type: React.ElementType<P>,
+  props?: P | null,
+  ...children: React.ReactNode[]
+): React.ReactElement<P> | null {
+  return isSlotComponent(props)
+    ? jsxFromSlotComponent(type, props, ...children)
+    : React.createElement(type, props, ...children);
+}
+
+function jsxFromSlotComponent<Props extends UnknownSlotProps>(
+  type: React.ElementType<Props>,
+  propsWithMetadata: WithMetadata<Props>,
+  ...overrideChildren: React.ReactNode[]
+): React.ReactElement<Props> | null {
+  const { children, renderFn, props } = normalizeChildren(propsWithMetadata, overrideChildren);
+  if (renderFn) {
+    return React.createElement(React.Fragment, {}, renderFn(type, { ...props, children })) as React.ReactElement<Props>;
+  }
+
+  return React.createElement<Props>(type, { ...props, children } as Props);
+}
+
+export function extractChildrenFromProps<Props extends React.PropsWithChildren<{}>>(
+  props?: Props | null,
+): React.ReactNode[] {
+  if (!props) {
+    return [];
+  }
+  return Array.isArray(props.children) ? props.children : [props.children];
+}
+
+function normalizeChildren<Props extends UnknownSlotProps>(
+  propsWithMetadata: WithMetadata<Props>,
+  overrideChildren: React.ReactNode[] = [],
+): {
+  props: Props;
+  children: React.ReactNode;
+  renderFn?: SlotRenderFunction<Props>;
+} {
+  const {
+    [SLOT_EXTERNAL_CHILDREN_SYMBOL]: slotExternalChildren,
+    [SLOT_INTERNAL_CHILDREN_SYMBOL]: slotInternalChildren,
+    ...props
+  } = propsWithMetadata;
+  const isRenderFn = typeof slotExternalChildren === 'function';
+  const notNormalizedChildren: React.ReactNode | React.ReactNode[] =
+    (overrideChildren.length > 0 ? overrideChildren : undefined) ??
+    (isRenderFn ? slotInternalChildren : slotExternalChildren) ??
+    slotInternalChildren ??
+    propsWithMetadata.children;
+  const normalizedChildren = Array.isArray(notNormalizedChildren)
+    ? React.createElement(React.Fragment, {}, ...notNormalizedChildren)
+    : notNormalizedChildren;
+  const renderFn = isRenderFn ? (slotExternalChildren as SlotRenderFunction<Props>) : undefined;
+  if (isRenderFn) {
+    const { children: _, ...propsWithoutChildren } = props as UnknownSlotProps;
+    return {
+      children: normalizedChildren,
+      renderFn,
+      props: propsWithoutChildren as Props,
+    };
+  }
+  return {
+    children: normalizedChildren,
+    renderFn,
+    props: props as UnknownSlotProps as Props,
+  };
```

Comment:
We care about the case with a render function only, correct? Should it be following instead (we don't need to normalize children, we have already them computed)?

```
  const {
    [SLOT_EXTERNAL_CHILDREN_SYMBOL]: slotExternalChildren,
    [SLOT_INTERNAL_CHILDREN_SYMBOL]: slotInternalChildren,
    ...props
  } = propsWithMetadata;

  const isRenderFn = typeof slotExternalChildren === 'function';


  if (isRenderFn) {
    // other logic
  }

  return {
    children: props.children,
    renderFn: undefined,
    props: props as UnknownSlotProps as Props,
  };
```

Reply:
With that if I am not missing a case, we can do following:

```ts
  if (isRenderFn) {
    const normalizedChildren: React.ReactNode | React.ReactNode[] =
      (overrideChildren.length > 0 ? overrideChildren : undefined) ??
      slotInternalChildren ??
      propsWithMetadata.children;
    const renderFn = isRenderFn ? (slotExternalChildren as SlotRenderFunction<Props>) : undefined;
    
    return {
      children: normalizedChildren,
      renderFn,
      props: props as UnknownSlotProps as Props,
    };
  }
```

(if no, let's add a unit test)